python/tchannel/tornado/peer.py

```diff
@@ -458,35 +492,95 @@ def send(self, arg1, arg2, arg3, traceflag=False, headers=None):
-                headers=headers,
-                tracing=Trace(
-                    name=endpoint,
-                    trace_id=trace_id,
-                    parent_span_id=parent_span_id,
-                    endpoint=Endpoint(self.peer.host,
-                                      self.peer.port,
-                                      self.service),
-                    traceflags=traceflag,
-                )
+        headers[THeaders.ARG_SCHEME] = headers.get(
+            THeaders.ARG_SCHEME,
+            ArgSchemeFlag.raw,
+        )
+        headers[THeaders.RETRY] = headers.get(
+            THeaders.RETRY,
+            RetryFlag.default,
+        )
+
+        original_request = Request(
+            service=self.service,
+            argstreams=[InMemStream(endpoint), arg2, arg3],
+            id=0,
+            headers=headers,
+            tracing=Trace(
+                name=endpoint,
+                trace_id=trace_id,
+                parent_span_id=parent_span_id,
+                endpoint=Endpoint(self.peer.host,
+                                  self.peer.port,
+                                  self.service),
+                traceflags=traceflag,
             )
         )
 
+        response = None
+        black_list = []
+        # only retry on non-stream request
+        if is_stream:
+            connection = yield self.peer.connect()
+            message_id = connection.next_message_id()
+            original_request.id = message_id
+            response = yield connection.send_request(original_request)
+        else:
+            req = original_request
+            start_time = time.time()
+            try:
+                peer = self.peer
+                # mac number of times to retry 5.
+                for i in range(Request.MAX_RETRY_TIMES):
+                    connection = yield peer.connect()
+                    message_id = connection.next_message_id()
+                    req = original_request.clone()
+                    req.id = message_id
+                    try:
+                        time_left = Request.TIMEOUT - (
+                            time.time() - start_time)
+                        # request timeout 5s
+                        response_future = connection.send_request(req)
+                        with timeout(response_future, time_left):
+                            response = yield response_future
+                        break
+                    except ProtocolError as e:
+                        if not should_retry_on_error(req, e):
+                            raise
+                        black_list.append(peer)
+                        self.kwargs['black_list'] = black_list
+                        peer = self.peer_group.choose(**self.kwargs)
+                        if not peer or i == Request.MAX_RETRY_TIMES - 1:
+                            raise
+            except TimeoutError:
+                req.set_exception(TimeoutError)
+                raise
+
```

Comment:
+1 - start by pulling the functionality inside the for loop into its own function.